beacon-chain/endtoend/endtoend_test.go

```diff
@@ -0,0 +1,514 @@
+	if !found {
+		t.Fatal("validator binary not found")
+	}
+
+	if validatorNum%beaconNodeNum != 0 {
+		t.Fatal("Validator count is not easily divisible by beacon node count.")
+	}
+
+	valClients := make([]*validatorClientInfo, beaconNodeNum)
+	validatorsPerNode := validatorNum / beaconNodeNum
+	for n := uint64(0); n < beaconNodeNum; n++ {
+		file, err := os.Create(path.Join(tmpPath, fmt.Sprintf("vals%d.log", n)))
+		if err != nil {
+			t.Fatal(err)
+		}
+		args := []string{
+			fmt.Sprintf("--interop-num-validators=%d", validatorsPerNode),
+			fmt.Sprintf("--interop-start-index=%d", validatorsPerNode*n),
+			fmt.Sprintf("--monitoring-port=%d", 9080+n),
+			fmt.Sprintf("--beacon-rpc-provider=localhost:%d", 4000+n),
+		}
+		cmd := exec.Command(binaryPath, args...)
+		cmd.Stdout = file
+		cmd.Stderr = file
+		if err := cmd.Start(); err != nil {
+			t.Fatal(err)
+		}
+		valClients[n] = &validatorClientInfo{
+			processID:   cmd.Process.Pid,
+			monitorPort: 9080 + n,
+		}
+	}
+
+	for n := uint64(0); n < beaconNodeNum; n++ {
+		file, err := os.Open(path.Join(tmpPath, fmt.Sprintf("vals%d.log", n)))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if err = WaitForTextInFile(file, "Waiting for beacon chain start log"); err != nil {
+			t.Fatal(err)
+		}
+
+		t.Logf("%d Validators started for beacon node %d", validatorsPerNode, n)
+	}
+
+	client, err := rpc.Dial(path.Join(tmpPath, "eth1data/geth.ipc"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	web3 := ethclient.NewClient(client)
+
+	jsonBytes, err := ioutil.ReadFile(keystorePath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	r := bytes.NewReader(jsonBytes)
+	txOps, err := bind.NewTransactor(r, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	minDeposit := big.NewInt(int64(params.BeaconConfig().MaxEffectiveBalance))
+	txOps.Value = minDeposit.Mul(minDeposit, big.NewInt(1e9))
+	txOps.GasLimit = 4000000
+
+	contract, err := contracts.NewDepositContract(contractAddress, web3)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	deposits, _, _ := testutil.SetupInitialDeposits(t, validatorNum)
+	var tx *types.Transaction
+	for _, dd := range deposits {
+		tx, err = contract.Deposit(txOps, dd.Data.PublicKey, dd.Data.WithdrawalCredentials, dd.Data.Signature)
+		if err != nil {
+			t.Error("unable to send transaction to contract")
+			continue
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	// Wait for last tx to mine.
+	for pending := true; pending; _, pending, err = web3.TransactionByHash(context.Background(), tx.Hash()) {
+		if err != nil {
+			t.Fatal(err)
+		}
+		time.Sleep(1 * time.Second)
+	}
+
+	keystore, err := keystore.DecryptKey(jsonBytes, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := MineBlocks(web3, keystore, 20); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Logf("%d deposits mined", len(deposits))
+	return valClients
+}
+
+func PeersConnect(port uint64, expectedPeers uint64) error {
+	response, err := http.Get(fmt.Sprintf("http://127.0.0.1:%d/p2p", port))
+	if err != nil {
+		return err
+	}
+	time.Sleep(2 * time.Second)
```

Comment:
Libp2p is super hard to test, if you can't think of a better way to do this, please just add a comment to explain the sleep.